### PR TITLE
Sync groups on login

### PR DESCRIPTION
**Pull request by Matthijsy**
Hi,

Our Custom Oauth provider has the feature of group management. It would be really great to sync those groups between our provider and nextcloud. I can imagine that more people will have this feature in their oauth provider so this is a suggestion to implement it.

What the code does is creating a group called `provider-id` and link the user to it. In the feature it would be great if it can be a custom name but that is something that needs to be implemented in nextcloud first.

I would like to get some feedback and will try to improve my code if necessary.

### CODE REVIEW

**Comment by zorn-v** [comment]
I prefer did not include "provider related" features. "hd" from google - and I got something like this https://github.com/zorn-v/nextcloud-social-login/issues/85
https://github.com/zorn-v/nextcloud-social-login/pull/82

But I check code

File: lib/Controller/LoginController.php [hunk 1]
```diff
@@ -71,6 +73,7 @@ public function __construct(
         $this->session = $session;
         $this->l = $l;
         $this->socialConnect = $socialConnect;
+        $this->provider = null;
```
Comment (zorn-v) [inline lib/Controller/LoginController.php]: What point of this ? It is already `null`

File: lib/Controller/LoginController.php [hunk 2]
```diff
@@ -328,6 +332,27 @@ private function login($uid, Profile $profile)
             }
         }
 
+        # Create groups via the provider
+        $group_ids = [];
+        foreach($profile->data->groups as $provider_group) {
```
Comment (zorn-v) [inline lib/Controller/LoginController.php]: If no `data->groups` in response will be warning

File: lib/Controller/LoginController.php [hunk 3]
```diff
@@ -244,6 +247,7 @@ public function telegram()
 
     private function auth($class, array $config, $provider, $providerTitle)
     {
+        $this->provider = $provider;
```
Comment (zorn-v) [inline lib/Controller/LoginController.php]: Bad design 😉

**Comment by zorn-v** [comment]
So, you write code only for yourself. Sorry but I can't merge it

File: lib/Controller/LoginController.php (hunk 2 above)
Comment (Matthijsy) [reply lib/Controller/LoginController.php]: Where will there be a warning? I didn't found a warning and everything worked fine even without data->groups.

File: lib/Controller/LoginController.php (hunk 3 above)
Comment (Matthijsy) [reply lib/Controller/LoginController.php]: Agreed, I did change it

**Comment by Matthijsy** [comment]
Thanks for your fast response. I didn't write it only for myself (then there is no point of opening a pr, could just keep it for myself). I did write it because we need it (but if that was the only thing I can keep it local and don't merge it back to the project itself). I did open this PR because I think this can be usefull for other people as well. 

Nextcloud supports LDAP with groups, but setting up LDAP is a lot of work so probably a lot of people don't want to do that. However OAuth is getting more popular so I expect more people wanting to use OAuth with NC (already an issue about that: https://github.com/nextcloud/server/issues/5955) and probably also wanting to sync there groups from their OAuth provider. This implementation will support that. 

So it's not only for myself but for the community to be able to sync their group system.

File: lib/Controller/LoginController.php (hunk 2 above)
Comment (zorn-v) [reply lib/Controller/LoginController.php]: Check logs. At least will be something like `Invalid argument supplied for foreach()` plus `Trying to get property 'groups' of non-object` if no `data` in response etc.

**Comment by zorn-v** [comment]
> so that this is not provider related but that every provider can use it

Ok, Which provider have "data->groups" in profile info ? 
Google, Facebook, GitHab, GitLab, Wordpress ?

What about support existing "add new users to specified group" setting ?

**Comment by funkyferdy** [comment]
Hello @zorn-v !
>>>Ok, Which provider have "data->groups" in profile info ?
WSO2 Identity Server :)
Please take a look at my Issue here: https://github.com/zorn-v/nextcloud-social-login/issues/87

**Comment by Matthijsy** [comment]
@funkyferdy Great to hear that you also have the need for group sync. I don't know WSO2 ID Server, but does it provide group names in the structure as described in the changed readme? (In the data->groups object?). 

You mentioned a kind of mapping between the groups in WSO2 and NextCloud. With my current implementation the groups will be copied 1-to-1 (so a group with id 'admin' in your provider with name 'my-provider' will be called 'my-provider-admin'). Will this work for your usecase? Or do you really need this kind of mapping (since this will gives a lot more complexity).

Would like to hear from you so that this could also work for more cases

**Comment by funkyferdy** [comment]
Hello @Matthijsy!

>>>but does it provide group names in the structure as described in the changed readme?

Yes, by default is a claim named "Role". For our NGO im working for we have 16k so called "Roles". In fact, WSO Identity Server is very flexible on that point. See also here (https://docs.wso2.com/display/IS570/Configuring+Claims+for+a+Service+Provider). I could rename the role Claim to whatever is needed from nextcloud side. 

But that is not the point. On application side (Nextcloud) it would be great to be able to be able to configure mainly 2 things:

1) Only Users with role "official" are able to login to nextcloud at all.
(Example: A "subcriber" is not allowed but a "official" is allowed to use nextcloud)

2) When it comes to "map" a Identity Provider "role" to a exiting nextcloud "Group" the usecase that i have in mind is to "put" the User in the right nextloud-Group automaticly. Example is following: Here comes a User that is allowed to login to nexcloud, he has role "official". Now, on Login/creation of the user in nexcloud it should be put also in Nextcloud-Group "Officials-Board" because he has also a Role "Officials-Board" in Identity Server.

BTW, we have developped exactly this functionality for TYPO3-CMS so time ago. Take a look at this repo if you like: https://github.com/xperseguers/t3ext-oidc/

**Comment by zorn-v** [comment]
> Example is following: Here comes a User that is allowed to login to nexcloud, he has role "official". Now, on Login/creation of the user in nexcloud it should be put also in Nextcloud-Group "Officials-Board" because he has also a Role "Officials-Board" in Identity Server.

Is there many roles (more than one) that can login ? Maybe this https://github.com/zorn-v/nextcloud-social-login/issues/88 feature is enough which I plan to implement in near future

**Comment by Matthijsy** [comment]
In my case there are multiple groups that need to be able to login (all of them so point 1 is not really relavant for my use case), and to be synced. I dont know about the WSO case.

**Comment by funkyferdy** [comment]
>>Is there many roles (more than one) that can login ? 

Potencially yes. Maybe you have Division 1 and Division 2 is allowed but not Division 3. Here configurable for a specific indentity provider. 

in A second step you want to configure for a specific Identity Provider the roles->Group assignment.

In my opinion this are 2 different features.

**Comment by fabianfreyer** [comment]
> So, you write code only for yourself. Sorry but I can't merge it

I'd also be very interested in this feature. We are also using a custom Oauth2 provider, and would love to sync groups with it. Ideally this would be with a group field on the profile page, or a separate endpoint that returns the list of groups.
We have around 10 groups with a few hundred users, and manually synchronizing these in multiple places is a hassle and error-prone.

**Comment by Matthijsy** [comment]
@zorn-v any thoughs about this PR? I think the need for having a group sync is there, I am wondering if this implementation is the way to go or that I should change something

**Comment by proietfb** [comment]
Hi guys.
Also we are looking for something to map our IAM user groups.
Currently we have had patched in our custom clone of this repo waiting to make PR. 
But also here you are :)

**Comment by zorn-v** [comment]
Implemented in v1.15.0
Check readme for details https://github.com/zorn-v/nextcloud-social-login#custom-oidc-groups

**Comment by fabianfreyer** [comment]
Thanks!

**Comment by Matthijsy** [comment]
@zorn-v How do I get this working with an Oauth provider? @fabianfreyer You also use a Oauth provider right? How did you get this to work? If I just move my settings from the Oauth settings to the Custom OpenID Connect settings it will not login anymore.

**Comment by zorn-v** [comment]
Which provider are you using ?

**Comment by zorn-v** [comment]
I added new settings (groups and logout url) to custom oauth in v1.16.1

**Comment by Matthijsy** [comment]
Thanks it works! Really great!

**Pull request closed without merging**